keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,402 @@
+  `PulledImage` and `FailedToPullImage` but these will contain additional
+  metadata, like the pull progress.
+
+- **Will enabling / using the feature result in increasing size or count of
+  the existing API objects?**
+
+  Yes, the amount of events will increase up to the configured event period
+  multiplied by the amount of pods pulling an image at the same time.
+
+- **Will enabling / using the feature result in increasing time taken by any
+  operations covered by existing SLIs/SLOs (e.g. by adding additional
+  work, introducing new steps in between, etc.)? Please describe the details
+  if so.**
+
+  No, it should not influence anything mentioned. The calculation of the
+  pull progress is left out for sake of triviality.
+
+- **Will enabling / using the feature result in non-negligible increase of
+  resource usage (CPU, RAM, disk IO, ...) in any components? Things to keep in
+  mind include: additional in-memory state, additional non-trivial
+  computations, excessive access to disks (including increased log volume),
+  significant amount of data sent and/or received over network, etc.**
+
+  No
+
+### Rollout, Upgrade, and Rollback Planning
+
+### Dependencies
+
+- **Does this feature depend on any specific services running in the cluster
+  (e.g., a metrics service)?**
+
+  Yes the container runtime needs to support the new streaming CRI API method
+  `PullImageProgress()`.
+
+- **How does this feature respond to complete failures of the services on
+  which it depends?**
+
+  If the feature does not work as expected or does not work at all, the
+  overall image pull functionality will not be affected, but the additional
+  metadata (pull progress) is not visible to the user.
+
+- **How does this feature respond to degraded performance or high error rates
+  from services on which it depends?**
+
+  It will report these errors to the API via events. This was not done before
+  and is also a purpose of this feature.
+
+### Monitoring requirements
+
+- **How can an operator determine if the feature is in use by workloads?**
+
+  Checking if its enabled in the kubelet configuration.
+
+- **How can an operator determine if the feature is functioning properly?**
+
+  Image pull progress should be exposed via events in the configured intervals.
+
+- **What are the service level indicators an operator can use to determine the
+  health of the service?**
+
+  The kubelet is reporting a prometheus metric for the number of failed calls
+  to the new CRI API method or failed event deliveries. The SLI will correlate
+  to the aggregation of this metric for all nodes.
+
+- **What are reasonable service level objectives for the feature?**
+
+  (I'm unsure about this question)
+
+### Troubleshooting
+
+- **What are the known failure modes?**
+
+  The will not start if the feature is enabled (configured) but not supported
+  by the runtime.
+
+- **How can those be detected via metrics or logs?**
+
+  Yes, the kubelet will log the failure.
```

Comment:
Any metric you're planning to add?

Reply:
I would add the metric to the container runtime. I don't think that the kubelet needs an additional metric for it, WDYT?

Reply:
Hm in terms of troubleshooting, I think an additional kubelet metric would make sense...